File: apps/agents/desktop/telemetry/sysfs_linux.go
//go:build linux

package telemetry

import (
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

// sysfsCache holds resolved sysfs paths so we don't re-glob every tick.
// Linux hwmon numbering is stable per boot.
type sysfsCache struct {
	once sync.Once

	cpuTempPath  string
	gpuBusyPath  string
	gpuTempPath  string
	gpuFreqPath  string
	gpuMemFreq   string
	vramUsedPath string
	vramTotal    string
	powerCapPath string
	powerAvgPath string
	fanPath      string
	batteryPath  string
	cpuFreqPaths []string
}

var paths sysfsCache

func (c *sysfsCache) resolve() {
	c.once.Do(func() {
		// CPU temperature: k10temp (AMD) or coretemp (Intel)
		hwmonDir := "/sys/class/hwmon"
		entries, err := os.ReadDir(hwmonDir)
		if err == nil {
			for _, entry := range entries {
				base := filepath.Join(hwmonDir, entry.Name())
				nameBytes, err := os.ReadFile(filepath.Join(base, "name"))
				if err != nil {
					continue
				}
				name := strings.TrimSpace(string(nameBytes))

				if (name == "k10temp" || name == "coretemp") && c.cpuTempPath == "" {
					c.cpuTempPath = filepath.Join(base, "temp1_input")
				}
				if fanPath := filepath.Join(base, "fan1_input"); fileExists(fanPath) && c.fanPath == "" {
					c.fanPath = fanPath
				}
				if capPath := filepath.Join(base, "power1_cap"); fileExists(capPath) && c.powerCapPath == "" {
					c.powerCapPath = capPath
				}
				for _, pName := range []string{"power1_average", "power1_input"} {
					pPath := filepath.Join(base, pName)
					if fileExists(pPath) && c.powerAvgPath == "" {
						c.powerAvgPath = pPath
						break
					}
				}
			}
		}

		// GPU paths (AMDGPU — card0/card1)
		cards, _ := filepath.Glob("/sys/class/drm/card[0-9]")
		for _, card := range cards {
			busy := filepath.Join(card, "device", "gpu_busy_percent")
			if !fileExists(busy) {
				continue
			}
			c.gpuBusyPath = busy

			hwmons, _ := filepath.Glob(filepath.Join(card, "device", "hwmon", "hwmon*"))
			for _, hwmon := range hwmons {
				temp := filepath.Join(hwmon, "temp1_input")
				if fileExists(temp) {
					c.gpuTempPath = temp
					break
				}
			}

			if freq := filepath.Join(card, "device", "pp_dpm_sclk"); fileExists(freq) {
				c.gpuFreqPath = freq
			}
			if mclk := filepath.Join(card, "device", "pp_dpm_mclk"); fileExists(mclk) {
				c.gpuMemFreq = mclk
			}
			if vt := filepath.Join(card, "device", "mem_info_vram_total"); fileExists(vt) {
				c.vramTotal = vt
			}
			if vu := filepath.Join(card, "device", "mem_info_vram_used"); fileExists(vu) {
				c.vramUsedPath = vu
			}
			break
		}

		// Battery
		bats, _ := filepath.Glob("/sys/class/power_supply/BAT*")
		if len(bats) > 0 {
			c.batteryPath = bats[0]
		}

		// CPU frequency paths
		c.cpuFreqPaths, _ = filepath.Glob("/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq")
	})
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

// readCPUTimes parses /proc/stat to get aggregate CPU idle and total jiffies.
func readCPUTimes() (idle, total uint64) {
	data, err := os.ReadFile("/proc/stat")
	if err != nil {
		return 0, 0
	}

	// First line: "cpu  user nice system idle iowait irq softirq steal ..."
	lines := strings.SplitN(string(data), "\n", 2)
	if len(lines) == 0 {
		return 0, 0
	}

	fields := strings.Fields(lines[0])
	if len(fields) < 5 || fields[0] != "cpu" {
		return 0, 0
	}

	var sum uint64
	for i := 1; i < len(fields); i++ {
		v, err := strconv.ParseUint(fields[i], 10, 64)
		if err != nil {
			continue
		}
		sum += v
		if i == 4 { // idle is the 4th value (index 4 after "cpu")
			idle = v
		}
	}

	return idle, sum
}

// calculateCPUUsage computes CPU usage percentage from two snapshots.
func calculateCPUUsage(prevIdle, prevTotal, currIdle, currTotal uint64) float64 {
	deltaTotal := currTotal - prevTotal
	deltaIdle := currIdle - prevIdle

	if deltaTotal == 0 {
		return 0
	}

	usage := (1.0 - float64(deltaIdle)/float64(deltaTotal)) * 100
	return math.Round(usage*10) / 10
}

// readCPUTemp reads CPU temperature in degrees C from cached hwmon path.
func readCPUTemp() float64 {
	paths.resolve()
	if paths.cpuTempPath == "" {
		return -1
	}

	tempBytes, err := os.ReadFile(paths.cpuTempPath)
	if err != nil {
		return -1
	}

	millideg, err := strconv.ParseInt(strings.TrimSpace(string(tempBytes)), 10, 64)
	if err != nil {
		return -1
	}

	return float64(millideg) / 1000.0
}

// readGPUUsage reads GPU busy percentage from cached sysfs path.
func readGPUUsage() float64 {
	paths.resolve()
	if paths.gpuBusyPath == "" {
		return -1
	}

	data, err := os.ReadFile(paths.gpuBusyPath)
	if err != nil {
		return -1
	}

	val, err := strconv.ParseFloat(strings.TrimSpace(string(data)), 64)
	if err != nil {
		return -1
	}

	return val
}

// readGPUTemp reads GPU temperature from cached hwmon path.
func readGPUTemp() float64 {
	paths.resolve()
	if paths.gpuTempPath == "" {
		return -1
	}

	data, err := os.ReadFile(paths.gpuTempPath)
	if err != nil {
		return -1
	}

	millideg, err := strconv.ParseInt(strings.TrimSpace(string(data)), 10, 64)
	if err != nil {
		return -1
	}

	return float64(millideg) / 1000.0
}

// readMemInfo parses /proc/meminfo to get total, available, swap total and swap free in bytes.
func readMemInfo() (total, available, swapTotal, swapFree int64) {
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return -1, -1, -1, -1
	}

	var gotTotal, gotAvailable, gotSwapTotal, gotSwapFree bool

	for _, line := range strings.Split(string(data), "\n") {
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}

		val, err := strconv.ParseInt(fields[1], 10, 64)
		if err != nil {
			continue
		}

		// Values in /proc/meminfo are in kB
		switch fields[0] {
		case "MemTotal:":
			total = val * 1024
			gotTotal = true
		case "MemAvailable:":
			available = val * 1024
			gotAvailable = true
		case "SwapTotal:":
			swapTotal = val * 1024
			gotSwapTotal = true
		case "SwapFree:":
			swapFree = val * 1024
			gotSwapFree = true
		}

		if gotTotal && gotAvailable && gotSwapTotal && gotSwapFree {
			break
		}
	}

	if !gotTotal {
		return -1, -1, -1, -1
	}
	if !gotSwapTotal {
		swapTotal = 0
	}
	if !gotSwapFree {
		swapFree = 0
	}

	return total, available, swapTotal, swapFree
}

// readBattery reads battery capacity and status from cached path.
func readBattery() (capacity int, status string) {
	paths.resolve()
	if paths.batteryPath == "" {
		return -1, ""
	}

	capBytes, err := os.ReadFile(filepath.Join(paths.batteryPath, "capacity"))
	if err != nil {
		return -1, ""
	}

	cap, err := strconv.Atoi(strings.TrimSpace(string(capBytes)))
	if err != nil {
		return -1, ""
	}

	statusBytes, err := os.ReadFile(filepath.Join(paths.batteryPath, "status"))
	if err != nil {
		return cap, "Unknown"
	}

	return cap, strings.TrimSpace(string(statusBytes))
}

// readCPUFreq returns the average CPU frequency across all cores in MHz.
func readCPUFreq() float64 {
	paths.resolve()
	if len(paths.cpuFreqPaths) == 0 {
		return -1
	}

	var total float64
	var count int
	for _, path := range paths.cpuFreqPaths {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		// Value is in kHz
		val, err := strconv.ParseFloat(strings.TrimSpace(string(data)), 64)
		if err != nil {
			continue
		}
		total += val
		count++
	}

	if count == 0 {
		return -1
	}

	return math.Round(total / float64(count) / 1000)
}

// readGPUFreq reads the current GPU frequency from pp_dpm_sclk (AMD) in MHz.
func readGPUFreq() float64 {
	paths.resolve()
	if paths.gpuFreqPath == "" {
		return -1
	}

	data, err := os.ReadFile(paths.gpuFreqPath)
	if err != nil {
		return -1
	}

	return parseDPMFreq(string(data))
}

// readPowerInfo reads TDP cap and current power draw from cached hwmon paths in watts.
func readPowerInfo() (tdpWatts, powerWatts float64) {
	paths.resolve()
	tdpWatts = -1
	powerWatts = -1

	if paths.powerCapPath != "" {
		if capData, err := os.ReadFile(paths.powerCapPath); err == nil {
			if val, err := strconv.ParseInt(strings.TrimSpace(string(capData)), 10, 64); err == nil {
				tdpWatts = float64(val) / 1000000.0
			}
		}
	}

	if paths.powerAvgPath != "" {
		if powerData, err := os.ReadFile(paths.powerAvgPath); err == nil {
			if val, err := strconv.ParseInt(strings.TrimSpace(string(powerData)), 10, 64); err == nil {
				powerWatts = float64(val) / 1000000.0
			}
		}
	}

	return tdpWatts, powerWatts
}

// readVRAMInfo reads VRAM used and total bytes from cached sysfs paths.
func readVRAMInfo() (used, total int64) {
	paths.resolve()
	if paths.vramTotal == "" {
		return -1, -1
	}

	totalData, err := os.ReadFile(paths.vramTotal)
	if err != nil {
		return -1, -1
	}
	totalVal, err := strconv.ParseInt(strings.TrimSpace(string(totalData)), 10, 64)
	if err != nil {
		return -1, -1
	}

	if paths.vramUsedPath == "" {
		return -1, totalVal
	}

	usedData, err := os.ReadFile(paths.vramUsedPath)
	if err != nil {
		return -1, totalVal
	}
	usedVal, err := strconv.ParseInt(strings.TrimSpace(string(usedData)), 10, 64)
	if err != nil {
		return -1, totalVal
	}

	return usedVal, totalVal
}

// readGPUMemFreq reads GPU memory clock from pp_dpm_mclk (AMD) in MHz.
func readGPUMemFreq() float64 {
	paths.resolve()
	if paths.gpuMemFreq == "" {
		return -1
	}

	data, err := os.ReadFile(paths.gpuMemFreq)
	if err != nil {
		return -1
	}

	return parseDPMFreq(string(data))
}

// parseDPMFreq extracts the active frequency from a pp_dpm_* file.
// Looks for the line marked with *, falls back to the last entry.
func parseDPMFreq(content string) float64 {
	var lastFreq float64 = -1

	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}

		freqStr := strings.TrimSuffix(fields[1], "Mhz")
		val, err := strconv.ParseFloat(freqStr, 64)
		if err != nil {
			continue
		}

		if strings.Contains(line, "*") {
			return val
		}
		lastFreq = val
	}

	return lastFreq
}

// readFanSpeed reads the first available fan speed in RPM from cached hwmon path.
func readFanSpeed() int {
	paths.resolve()
	if paths.fanPath == "" {
		return -1
	}

	data, err := os.ReadFile(paths.fanPath)
	if err != nil {
		return -1
	}

	if val, err := strconv.Atoi(strings.TrimSpace(string(data))); err == nil {
		return val
	}

	return -1
}

File: crates/telemetry/src/sysfs_linux.rs
//! Linux sysfs/procfs metric readers.
//!
//! Paths are resolved once on first access and cached for the lifetime of
//! the process (they are stable per boot on Linux).

use std::path::{Path, PathBuf};
use std::sync::OnceLock;

// ---------------------------------------------------------------------------
// Path cache (resolved once via OnceLock)
// ---------------------------------------------------------------------------

struct SysfsPaths {
    cpu_temp: Option<PathBuf>,
    cpu_freq_paths: Vec<PathBuf>,
    gpu_busy: Option<PathBuf>,
    gpu_temp: Option<PathBuf>,
    gpu_freq: Option<PathBuf>,
    gpu_mem_freq: Option<PathBuf>,
    vram_used: Option<PathBuf>,
    vram_total: Option<PathBuf>,
    power_cap: Option<PathBuf>,
    power_avg: Option<PathBuf>,
    fan: Option<PathBuf>,
    battery: Option<PathBuf>,
}

static PATHS: OnceLock<SysfsPaths> = OnceLock::new();

fn cached_paths() -> &'static SysfsPaths {
    PATHS.get_or_init(resolve_paths)
}

fn resolve_paths() -> SysfsPaths {
    let mut paths = SysfsPaths {
        cpu_temp: None,
        cpu_freq_paths: Vec::new(),
        gpu_busy: None,
        gpu_temp: None,
        gpu_freq: None,
        gpu_mem_freq: None,
        vram_used: None,
        vram_total: None,
        power_cap: None,
        power_avg: None,
        fan: None,
        battery: None,
    };

    // Scan hwmon devices for CPU temp, power, and fan.
    if let Ok(entries) = std::fs::read_dir("/sys/class/hwmon") {
        for entry in entries.flatten() {
            let dir = entry.path();
            let name = read_trimmed(&dir.join("name")).unwrap_or_default();

            // CPU sensor (AMD k10temp or Intel coretemp).
            if name == "k10temp" || name == "coretemp" {
                if paths.cpu_temp.is_none() {
                    let p = dir.join("temp1_input");
                    if p.exists() {
                        paths.cpu_temp = Some(p);
                    }
                }
                if paths.power_cap.is_none() {
                    let p = dir.join("power1_cap");
                    if p.exists() {
                        paths.power_cap = Some(p);
                    }
                }
                if paths.power_avg.is_none() {
                    let p = dir.join("power1_average");
                    if p.exists() {
                        paths.power_avg = Some(p);
                    } else {
                        let p = dir.join("power1_input");
                        if p.exists() {
                            paths.power_avg = Some(p);
                        }
                    }
                }
            }

            // Fan sensor.
            if paths.fan.is_none() {
                let p = dir.join("fan1_input");
                if p.exists() {
                    paths.fan = Some(p);
                }
            }
        }
    }

    // GPU (AMD DRM card).
    if let Ok(entries) = std::fs::read_dir("/sys/class/drm") {
        let mut cards: Vec<_> = entries
            .flatten()
            .filter(|e| {
                e.file_name()
                    .to_str()
                    .is_some_and(|n| n.starts_with("card") && n.len() <= 6)
            })
            .collect();
        cards.sort_by_key(|e| e.file_name());

        for card in cards {
            let dev = card.path().join("device");
            let busy = dev.join("gpu_busy_percent");
            if busy.exists() {
                paths.gpu_busy = Some(busy);

                // GPU temp: device/hwmon/hwmon*/temp1_input
                if let Ok(hwmon_entries) = std::fs::read_dir(dev.join("hwmon")) {
                    for he in hwmon_entries.flatten() {
                        let p = he.path().join("temp1_input");
                        if p.exists() {
                            paths.gpu_temp = Some(p);
                            break;
                        }
                    }
                }

                let sclk = dev.join("pp_dpm_sclk");
                if sclk.exists() {
                    paths.gpu_freq = Some(sclk);
                }
                let mclk = dev.join("pp_dpm_mclk");
                if mclk.exists() {
                    paths.gpu_mem_freq = Some(mclk);
                }
                let vram_used = dev.join("mem_info_vram_used");
                if vram_used.exists() {
                    paths.vram_used = Some(vram_used);
                }
                let vram_total = dev.join("mem_info_vram_total");
                if vram_total.exists() {
                    paths.vram_total = Some(vram_total);
                }

                break; // Use first card with GPU busy.
            }
        }
    }

    // Battery.
    if let Ok(entries) = std::fs::read_dir("/sys/class/power_supply") {
        for entry in entries.flatten() {
            let name = entry.file_name();
            if name.to_str().is_some_and(|n| n.starts_with("BAT")) {
                paths.battery = Some(entry.path());
                break;
            }
        }
    }

    // CPU frequency paths.
    if let Ok(entries) = std::fs::read_dir("/sys/devices/system/cpu") {
        let mut freq_paths: Vec<PathBuf> = entries
            .flatten()
            .filter(|e| {
                e.file_name().to_str().is_some_and(|n| {
                    n.starts_with("cpu") && n[3..].chars().all(|c| c.is_ascii_digit())
                })
            })
            .map(|e| e.path().join("cpufreq/scaling_cur_freq"))
            .filter(|p| p.exists())
            .collect();
        freq_paths.sort();
        paths.cpu_freq_paths = freq_paths;
    }

    paths
}

// ---------------------------------------------------------------------------
// Public metric readers
// ---------------------------------------------------------------------------

/// Reads CPU idle and total jiffies from `/proc/stat`.
pub fn read_cpu_times() -> (u64, u64) {
    let content = match std::fs::read_to_string("/proc/stat") {
        Ok(c) => c,
        Err(_) => return (0, 0),
    };

    // First line: "cpu  user nice system idle iowait irq softirq steal ..."
    let line = match content.lines().next() {
        Some(l) if l.starts_with("cpu ") => l,
        _ => return (0, 0),
    };

    let fields: Vec<u64> = line
        .split_whitespace()
        .skip(1) // skip "cpu"
        .filter_map(|f| f.parse().ok())
        .collect();

    if fields.len() < 4 {
        return (0, 0);
    }

    let idle = fields[3]; // 4th field is idle.
    let total: u64 = fields.iter().sum();
    (idle, total)
}

/// Reads CPU temperature in Celsius.
pub fn read_cpu_temp() -> f64 {
    cached_paths()
        .cpu_temp
        .as_ref()
        .and_then(|p| read_i64(p))
        .map(|v| v as f64 / 1000.0)
        .unwrap_or(-1.0)
}

/// Reads average CPU frequency in MHz.
pub fn read_cpu_freq() -> f64 {
    let paths = &cached_paths().cpu_freq_paths;
    if paths.is_empty() {
        return -1.0;
    }

    let mut sum: u64 = 0;
    let mut count: u64 = 0;
    for p in paths {
        if let Some(khz) = read_i64(p) {
            sum += khz as u64;
            count += 1;
        }
    }

    if count == 0 {
        return -1.0;
    }

    (sum / count / 1000) as f64 // kHz → MHz, truncated.
}

/// Reads GPU busy percentage.
pub fn read_gpu_usage() -> f64 {
    cached_paths()
        .gpu_busy
        .as_ref()
        .and_then(|p| read_f64(p))
        .unwrap_or(-1.0)
}

/// Reads GPU temperature in Celsius.
pub fn read_gpu_temp() -> f64 {
    cached_paths()
        .gpu_temp
        .as_ref()
        .and_then(|p| read_i64(p))
        .map(|v| v as f64 / 1000.0)
        .unwrap_or(-1.0)
}

/// Reads GPU core clock from `pp_dpm_sclk`.
pub fn read_gpu_freq() -> f64 {
    cached_paths()
        .gpu_freq
        .as_ref()
        .and_then(|p| parse_dpm_freq(p))
        .unwrap_or(-1.0)
}

/// Reads GPU memory clock from `pp_dpm_mclk`.
pub fn read_gpu_mem_freq() -> f64 {
    cached_paths()
        .gpu_mem_freq
        .as_ref()
        .and_then(|p| parse_dpm_freq(p))
        .unwrap_or(-1.0)
}

/// Reads VRAM (used_bytes, total_bytes). Returns (-1, -1) if unavailable.
pub fn read_vram() -> (i64, i64) {
    let paths = cached_paths();
    let total = paths
        .vram_total
        .as_ref()
        .and_then(|p| read_i64(p))
        .unwrap_or(-1);

    if total < 0 {
        return (-1, -1);
    }

    let used = paths
        .vram_used
        .as_ref()
        .and_then(|p| read_i64(p))
        .unwrap_or(-1);

    (used, total)
}

/// Reads memory info from `/proc/meminfo`.
/// Returns (total, available, swap_total, swap_free) in bytes.
pub fn read_mem_info() -> (i64, i64, i64, i64) {
    let content = match std::fs::read_to_string("/proc/meminfo") {
        Ok(c) => c,
        Err(_) => return (-1, -1, -1, -1),
    };

    let mut total: i64 = -1;
    let mut available: i64 = -1;
    let mut swap_total: i64 = 0;
    let mut swap_free: i64 = 0;

    for line in content.lines() {
        if let Some(val) = parse_meminfo_kb(line, "MemTotal:") {
            total = val * 1024;
        } else if let Some(val) = parse_meminfo_kb(line, "MemAvailable:") {
            available = val * 1024;
        } else if let Some(val) = parse_meminfo_kb(line, "SwapTotal:") {
            swap_total = val * 1024;
        } else if let Some(val) = parse_meminfo_kb(line, "SwapFree:") {
            swap_free = val * 1024;
        }
    }

    (total, available, swap_total, swap_free)
}

/// Reads battery capacity (0-100) and status string.
pub fn read_battery() -> (i32, String) {
    let bat_path = match &cached_paths().battery {
        Some(p) => p,
        None => return (-1, String::new()),
    };

    let capacity = read_trimmed(&bat_path.join("capacity"))
        .and_then(|s| s.parse::<i32>().ok())
        .unwrap_or(-1);

    if capacity < 0 {
        return (-1, String::new());
    }

    let status = read_trimmed(&bat_path.join("status")).unwrap_or_else(|| "Unknown".into());

    (capacity, status)
}

/// Reads TDP (watts) and power draw (watts).
pub fn read_power_info() -> (f64, f64) {
    let paths = cached_paths();

    let tdp = paths
        .power_cap
        .as_ref()
        .and_then(|p| read_i64(p))
        .map(|v| v as f64 / 1_000_000.0)
        .unwrap_or(-1.0);

    let power = paths
        .power_avg
        .as_ref()
        .and_then(|p| read_i64(p))
        .map(|v| v as f64 / 1_000_000.0)
        .unwrap_or(-1.0);

    (tdp, power)
}

/// Reads fan speed in RPM.
pub fn read_fan_speed() -> i32 {
    cached_paths()
        .fan
        .as_ref()
        .and_then(|p| read_i64(p))
        .map(|v| v as i32)
        .unwrap_or(-1)
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Reads a file and trims whitespace.
fn read_trimmed(path: &Path) -> Option<String> {
    std::fs::read_to_string(path)
        .ok()
        .map(|s| s.trim().to_string())
}

/// Reads a file as i64.
fn read_i64(path: &Path) -> Option<i64> {
    read_trimmed(path).and_then(|s| s.parse().ok())
}

/// Reads a file as f64.
fn read_f64(path: &Path) -> Option<f64> {
    read_trimmed(path).and_then(|s| s.parse().ok())
}

/// Parses a `pp_dpm_sclk`/`pp_dpm_mclk` file to extract the active frequency.
///
/// Format:
/// ```text
/// 0: 400Mhz
/// 1: 1000Mhz
/// 2: 2500Mhz *
/// ```
/// The line with `*` is the current frequency. Falls back to the last entry.
fn parse_dpm_freq(path: &Path) -> Option<f64> {
    let content = std::fs::read_to_string(path).ok()?;
    let mut last_freq: Option<f64> = None;

    for line in content.lines() {
        let line = line.trim();
        // Extract freq from "N: 1234Mhz" or "N: 1234Mhz *"
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            let freq_str = parts[1].trim_end_matches("Mhz").trim_end_matches("MHz");
            if let Ok(freq) = freq_str.parse::<f64>() {
                last_freq = Some(freq);
                if line.contains('*') {
                    return Some(freq);
                }
            }
        }
    }

    last_freq
}

/// Parses a line from `/proc/meminfo` matching a prefix, returns value in kB.
fn parse_meminfo_kb(line: &str, prefix: &str) -> Option<i64> {
    if !line.starts_with(prefix) {
        return None;
    }
    line[prefix.len()..]
        .split_whitespace()
        .next()
        .and_then(|v| v.parse().ok())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn read_cpu_times_returns_values() {
        let (idle, total) = read_cpu_times();
        // On any Linux system, these should be > 0.
        assert!(total > 0, "total CPU jiffies should be > 0");
        assert!(idle > 0, "idle CPU jiffies should be > 0");
        assert!(idle <= total, "idle should be <= total");
    }

    #[test]
    fn read_mem_info_returns_values() {
        let (total, avail, swap_total, swap_free) = read_mem_info();
        assert!(total > 0, "total memory should be > 0");
        assert!(avail > 0, "available memory should be > 0");
        assert!(avail <= total, "available should be <= total");
        assert!(swap_total >= 0);
        assert!(swap_free >= 0);
    }

    #[test]
    fn parse_dpm_freq_active_marker() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("pp_dpm_sclk");
        std::fs::write(&path, "0: 400Mhz\n1: 1000Mhz\n2: 2500Mhz *\n").unwrap();

        assert_eq!(parse_dpm_freq(&path), Some(2500.0));
    }

    #[test]
    fn parse_dpm_freq_no_marker_uses_last() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("pp_dpm_sclk");
        std::fs::write(&path, "0: 400Mhz\n1: 1000Mhz\n2: 2500Mhz\n").unwrap();

        assert_eq!(parse_dpm_freq(&path), Some(2500.0));
    }

    #[test]
    fn parse_meminfo_kb_valid() {
        assert_eq!(
            parse_meminfo_kb("MemTotal:       16000000 kB", "MemTotal:"),
            Some(16000000)
        );
    }

    #[test]
    fn parse_meminfo_kb_mismatch() {
        assert_eq!(
            parse_meminfo_kb("MemAvailable:   8000000 kB", "MemTotal:"),
            None
        );
    }
}
